File: exercism/haskell/strain/Strain.hs
module Strain where

keep :: (a -> Bool) -> [a] -> [a]
keep _ [] = []
keep p (x:xs)
  | p x = x : rest
  | otherwise = rest
  where rest = keep p xs

discard :: (a -> Bool) -> [a] -> [a]
discard p = keep (not . p)

File: exercism/kotlin/strain/src/main/kotlin/Strain.kt
object Strain {
  fun<T> keep(items: Iterable<T>, p: (T) -> Boolean): List<T> {

    val result: MutableList<T> = arrayListOf()

    for (item in items) {
      if (p(item)) result += item
    }

    return result
  }

  fun<T> discard(items: Iterable<T>, p: (T) -> Boolean): List<T> =
    keep(items, { !p(it) })
}
